File: 0389-find-the-difference/0389-find-the-difference.py
# Aug 06, 2026, 07:44
# I'll make a string value approach (typecasting)

class Solution:
    def findTheDifference(self, s: str, t: str) -> str:
        s_value = 0
        t_value = 0
        # get the sum of the ASCII values of s
        for char in s:
            s_value += ord(char)
        
        # get the sum of the ASCII values of t
        for char in t:
            t_value += ord(char)
        
        # the difference of the two should return the value of 
        # the added_letter
        # now we just need to typecast it back into a char
        return chr(t_value - s_value)

File: 0389-find-the-difference/0389-find-the-difference.c
// Aug 06, 2026 15:45
// trying this problem again, but in C
// why? I don't want to get cooked for this semester's C 
// lab courses using C as the langauge medium
// hashmap frequency approach (via an array with the index representing the letter)

char findTheDifference(char* s, char* t) {
    // we use a frequency array of size 26 for the lowercase letters as a hashmap
    int counts[26] = {0};

    // we count the frequencies of characters in s
    for (int i = 0; s[i] != '\0'; i++) {
        counts[s[i] - 'a']++;
    }

    // after that, we decrement the frequencies
    // for each chracter in t
    for (int i = 0; t[i] != '\0'; i++) {
        counts[t[i] - 'a']--;
    }

    // then we find the character with a non-zero count
    for (int i = 0; i < 26; i++) {
        if (counts[i] != 0) {
            // typecast the ASCII value back to a char
            // since that is what we need to return
            return (char)(i + 'a');
        }
    }

    // if the loop finishes without returning a char
    // we return null (not NULL, since that represents a null pointer)
    return '\0';
}